### ISSUE TITLE
coord: Source names that use catalog object names don't work

### ISSUE THREAD

**Issue by jkosh44** [MaterializeInc/materialize#13682]
### What version of Materialize are you using?

v0.26.1-dev

### How did you install Materialize?

Built from source

### What is the issue?

If you create a source with the same name as a catalog object, then you cannot properly create a materialized view on that source. 
It's possible that this issue can manifest in different ways, I haven't done extensive testing.

#### Postgres
```
joe=# CREATE TABLE v ( a int);
CREATE TABLE
joe=# ALTER TABLE v REPLICA IDENTITY FULL;
ALTER TABLE
joe=# CREATE PUBLICATION mz_objects FOR TABLE v;
CREATE PUBLICATION
```

#### Materialize
```
materialize=> CREATE CONNECTION pgconn FOR POSTGRES HOST 'localhost', PORT 5432, USER joe, DATABASE joe;
CREATE CONNECTION
materialize=> CREATE SOURCE mz_objects FROM POSTGRES CONNECTION pgconn PUBLICATION 'mz_objects';
CREATE SOURCE
materialize=> CREATE MATERIALIZED VIEWS FROM SOURCE mz_objects (v);
ERROR:  source "mz_objects" does not exist

```

**Comment by benesch** [comment]
Wait, why doesn't this work? Strange. The user created source should shadow the mz_catalog schema. Maybe we have a bug in the search paths.

**Comment by jkosh44** [comment]
Ahh, so I have the expected behavior backwards, but there still seems to be an issue. Not quite sure what the cause it, I'll update the issue description.

**Comment by bkirwi** [comment]
I spent some time poking at this one this afternoon. I believe this is the relevant search path code: https://github.com/MaterializeInc/materialize/blob/6d7b6230100a397c6c88ad3744ff1365a5df04f0/src/adapter/src/catalog.rs#L1005-L1036

The resulting search path is: the temp schema, mz catalog, pg catalog, everything else. The builtin `mz_objects` is in the mz catalog, and our new `mz_objects` is in "everything else". Which I think explains the behaviour.

I'm a bit nervous to just go shuffling the search path to put user stuff first, since that seems like it could break a lot of other things. Is the desired / expected behaviour documented anywhere? (Or how would I go about convincing myself this is safe?)

Also, the error message is incorrect: the problem is not that `mz_objects` does not exist, but that the `mz_objects` entry we _did_ find has the wrong type. Here's the relevant code: https://github.com/MaterializeInc/materialize/blob/6d7b6230100a397c6c88ad3744ff1365a5df04f0/src/adapter/src/catalog.rs#L1258-L1263

I propose renaming the `UnknownSource` to something like `NotASource` and updating the display implementation, to give a more accurate error message to the user.

**Comment by jkosh44** [comment]
In general we try and mimic the behavior that PostgreSQL has. They don't have sources but I think using views is a good analogy.
```
postgres=# SELECT * FROM pg_authid LIMIT 1;
 oid | rolname  | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolbypassrls | rolconnlimit | rolpassword | rolvaliduntil 
-----+----------+----------+------------+---------------+-------------+-------------+----------------+--------------+--------------+-------------+---------------
  10 | postgres | t        | t          | t             | t           | t           | t              | t            |           -1 |             | 
(1 row)
postgres=# CREATE VIEW pg_authid AS SELECT 1;
CREATE VIEW
postgres=# SELECT * FROM pg_authid LIMIT 1;
 oid | rolname  | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolbypassrls | rolconnlimit | rolpassword | rolvaliduntil 
-----+----------+----------+------------+---------------+-------------+-------------+----------------+--------------+--------------+-------------+---------------
  10 | postgres | t        | t          | t             | t           | t           | t              | t            |           -1 |             | 
(1 row)

postgres=# SELECT * FROM public.pg_authid LIMIT 1;
 ?column? 
----------
        1
(1 row)

postgres=# DROP VIEW pg_authid;
ERROR:  "pg_authid" is not a view
HINT:  Use DROP TABLE to remove a table.
postgres=# DROP VIEW public.pg_authid;
DROP VIEW
```

So I think your suggestion of `NotASource` is a good one because that seems to be the same behavior as Postgres. Just make sure that you can use the source if you prepend the schema to the source name.

We also have a similar error for dropping things:
```
materialize=> CREATE TABLE t (a int);
CREATE TABLE
materialize=> DROP VIEW t;
ERROR:  materialize.public.t is not of type VIEW
materialize=> DROP SOURCE t;
ERROR:  materialize.public.t is not of type SOURCE
```
So we might want to try and keep the errors somewhat consistent.

**Comment by bkirwi** [comment]
Opened a PR for the error message improvement.

The core issue remains -- what to do about the search path, if anything. Looking at the [postgres docs](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PRIV) it seems like our current behaviour is closest to their behaviour... which is to say, they put the builtin table at the beginning of the search path unless specified otherwise. Assuming that we want to copy the upstream behaviour here, maybe there's nothing to do?

**Comment by bkirwi** [comment]
Alright, the error message fix has landed: the user will now get an "unexpected type" error instead of the incorrect "does not exist."

I propose to do nothing about the search path, since postgres pretty clearly documents that system tables shadow user tables, and IIUC we follow postgres on this sort of thing. However, I do plan wrap some tests around this behaviour to more clearly document it / catch any regressions. (cc @benesch, since you too found the current behaviour surprising... please let us know if that sounds like the wrong path!)

**Comment by benesch** [comment]
> Alright, the error message fix has landed: the user will now get an "unexpected type" error instead of the incorrect "does not exist."

🎉 

> I propose to do nothing about the search path, since postgres pretty clearly documents that system tables shadow user tables, and IIUC we follow postgres on this sort of thing. However, I do plan wrap some tests around this behaviour to more clearly document it / catch any regressions. (cc @benesch, since you too found the current behaviour surprising... please let us know if that sounds like the wrong path!)

Oh, right. I'm out of date here. We used to report the wrong search path (#6754) for our behavior. But that got fixed in #12222. I think we're all good here!
